### PR TITLE
Core: V3 Metadata Upgrade Validation and Testing

### PR DESCRIPTION
**Pull request by leangjonathan**
Fixes https://github.com/apache/iceberg/issues/10763.

Validates metadata format upgrade paths:
- (OK) V1 -> V2
- (OK) V2 -> V3
- (OK) V1 -> V3
  - internally represented as V1->V2->V3

Adds logic for multi-version upgrades to internally execute as a stepwise upgrade so that non-incremental upgrade paths do not need to be explicitly maintained

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 1]
```diff
@@ -1015,6 +1015,11 @@ public Builder upgradeFormatVersion(int newFormatVersion) {
           "Cannot downgrade v%s table to v%s",
           formatVersion,
           newFormatVersion);
+      Preconditions.checkArgument(
+          newFormatVersion <= formatVersion + 1,
+          "Cannot skip format version(s) to upgrade v%s table to v%s",
+          formatVersion,
+          newFormatVersion);
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1022]: Curious, what's the rationale behind preventing upgrades from 1 to 3 (or more generally skipping versions)? 

I think it should technically possible to enable that upgrade safely (let me know if I'm missing something there). Also, with this check if  someone who really wants to "skip", they would just upgrade to 2 and then right to 3 after that since there's no real additional logic. Subsequent writes would write out the new format's metadata after that.

For what it's worth, I don't think it'll be common so maybe this check doesn't really matter in practice.

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1022]: I guess we could allow this and just internally do the upgrade to 2 first? My only goal here would be to make sure that we don't support 2 different upgrade paths. 

` 1 -> 3` and `1 -> 2 -> 3`

But if upgrading 1 ->3 is just syntatic sugar for 1 -> 2 -> 3 I don't see the issue

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 2]
```diff
@@ -28,51 +28,79 @@
 public class TestFormatVersions extends TestBase {
   @Parameters(name = "formatVersion = {0}")
   protected static List<Object> parameters() {
-    return Arrays.asList(1);
+    return Arrays.asList(1, 2);
   }
 
   @TestTemplate
   public void testDefaultFormatVersion() {
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:35]: This seems wrong to me? Or at least it looks wrong, why would testDefaultFormatVersion by between 1 and 2?

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java [hunk 3]
```diff
@@ -1452,52 +1452,148 @@ public void testCreateV2MetadataThroughTableProperty() {
...
-  public void testUpgradeV1MetadataToV2ThroughTableProperty() {
-    Schema schema = new Schema(Types.NestedField.required(10, "x", Types.StringType.get()));
-
-    TableMetadata meta =
-        TableMetadata.newTableMetadata(
-            schema,
-            PartitionSpec.unpartitioned(),
-            null,
-            ImmutableMap.of(TableProperties.FORMAT_VERSION, "1", "key", "val"));
+  public void testReplaceMetadataThroughTablePropertySkipVersion() {
```
Comment (RussellSpitzer) [inline core/src/test/java/org/apache/iceberg/TestTableMetadata.java:1488]: I think this probably (for forward looking things) shouldn't be doing a single skip but should be testing all upgrades to the newest one or if we want to keep the combinatorial explosion to a minimum just upgrades to the latest version.

Ie if we had 5 versions we could either do all the versions:
```
1 -> 2
1 -> 3
1 -> 4
1 -> 5
2 -> 3
2 -> 4
...
4 -> 5
```
Or just do the jump to the latest:
```
1 -> 5
2 -> 5
3 -> 5
4 -> 5
```

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 1 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1022]: Ok looking through the rest of this update I think it probably shouldn't be disallowed to skip versions, we should just be internally doing the stepwise upgrade.

Because of that I think we need to remove all the gates on skipping versions but also all of the added tests should not assume that the "version skip" is just a single version, and instead just test skipping to latest. 

So basically my thoughts for testing strategy is
Test
```
N => N+ 1
And
N => Max_Supported Version
```
Thoughts?

File: core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java [hunk 4]
```diff
@@ -199,7 +199,21 @@ public void upgradeFormatVersion() {
   }
 
   @Test
-  public void upgradeFormatVersionForView() {
+  public void upgradeFormatVersionV3() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java:202]: so technically you don't need to add a new test. I'd probably just slightly adjust `upgradeFormatVersion()` and make it a parameterized test:
```
@ParameterizedTest
@ValueSource(ints = {2, 3}
public void upgradeFormatVersionTo(int version) {...}
```

I'd then do the same pattern for `upgradeFormatVersionForView()`:

```
@ParameterizedTest
@ValueSource(ints = {2, 3}
public void upgradeFormatVersionForViewTo(int version) {...}
```

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java [hunk 5]
```diff
@@ -1452,52 +1452,148 @@ public void testCreateV2MetadataThroughTableProperty() {
...
+                    baseTableVersion, newTableVersion));
+      }
+    }
+  }
 
-    meta =
-        meta.replaceProperties(
-            ImmutableMap.of(TableProperties.FORMAT_VERSION, "2", "key2", "val2"));
+  @Test
+  public void testUpgradeMetadataThroughTableProperty() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestTableMetadata.java:1531]: this sounds to me like this should also be a parameterized test, since you're having two loops. Might be worth checking out https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 1 above)
Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1022]: > My only goal here would be to make sure that we don't support 2 different upgrade paths.

Thanks for this clarification, that makes sense to me. Having a single upgrade path makes it easier to rationalize about any issues that come up for an update (we don't have to think about "Oh does X happen when it's 1 to 3 directly or something else". It's a single path and so in case of any issues it must be at some specific point in the path). 

I think I'm overall good with that, as long as we can keep the combinatorics to a minimum for testing. Which I think your strategy is as good as it gets?

Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1022]: > My only goal here would be to make sure that we don't support 2 different upgrade paths.

Thanks for this clarification, that makes sense to me. Having a single upgrade path makes it easier to rationalize about any issues that come up for an update (we don't have to think about "Oh does X happen when it's 1 to 3 directly or something else". It's a single path and so in case of any issues it must be at some specific point in the path). 

I think I'm overall good with that, as long as we can keep the combinatorics to a minimum for testing. Which I think your strategy is as good as it gets?

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 2 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:35]: I think this test is not named well. `table.ops().current().formatVersion()` is parameterized on the list of versions from above. I don't think this test is adding anything so we can just remove it

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 1 above)
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1022]: This sounds alright, I'll remove the assert here and find a way to perform a serial upgrade internally.

Some implementation notes I have so far:
- Currently the internal implementation of performing a format version upgrade is to add to the list of metadata changes a MetadataUpdate.UpgradeFormatVersion instance.
- `MetadataUpdate.UpgradeFormatVersion` tracks the target version but not the origin version. This may mean that multiple instances of MetadataUpdate.UpgradeFormatVersion should be populated into the list of metadata changes to simulate this multi-version upgrade.
- `UpdateRequirements` class which defines all upgrade requirements has no requirements for the `MetadataUpdate.UpgradeFormatVersion` class due to the fact that backwards compatibility has never been broken (yet).

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java (hunk 3 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestTableMetadata.java:1488]: Combinatorial explosion for n choose 2 is small for reasonable values of n < 100. Modifying the upgrade test to check all potential upgrade paths

File: core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java (hunk 4 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java:202]: Thanks updating this with parameterization

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java (hunk 5 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestTableMetadata.java]: Added parameterization via a Streamed argument provider

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 2 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java]: Removed the test

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 1 above)
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java]: Also added additional asserts in TestFormatVersions to check that appropriate `MetadataUpdate.UpgradeFormatVersion` instances are registered in changes correctly.

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 2 above)
Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java]: the test should be added back and updated to 
```
@TestTemplate
  public void testDefaultFormatVersion() {
    assertThat(table.ops().current().formatVersion()).isEqualTo(formatVersion);
  }
```

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java [hunk 6]
```diff
@@ -1451,50 +1457,67 @@ public void testCreateV2MetadataThroughTableProperty() {
...
+        .flatMap(
+            baseTableVersion ->
+                IntStream.rangeClosed(
+                        baseTableVersion + 1, TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION)
+                    .mapToObj(newTableVersion -> arguments(baseTableVersion, newTableVersion)));
+  }
+
+  @ParameterizedTest
+  @MethodSource("upgradeTableVersionProvider")
+  public void testReplaceMetadataThroughTableProperty(int baseTableVersion, int newTableVersion) {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestTableMetadata.java:1473]:
```suggestion
  public void testReplaceMetadataThroughTableProperty(int baseFormatVersion, int newFormatVersion) {
```

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java [hunk 7]
```diff
@@ -1451,50 +1457,67 @@ public void testCreateV2MetadataThroughTableProperty() {
...
         .containsEntry("key", "val")
         .containsEntry("key2", "val2")
         .doesNotContainKey(TableProperties.FORMAT_VERSION);
   }
 
-  @Test
-  public void testUpgradeV1MetadataToV2ThroughTableProperty() {
+  @ParameterizedTest
+  @MethodSource("upgradeTableVersionProvider")
+  public void testUpgradeMetadataThroughTableProperty(int baseTableVersion, int newTableVersion) {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestTableMetadata.java:1502]:
```suggestion
  public void testUpgradeMetadataThroughTableProperty(int baseFormatVersion, int newFormatVersion) {
```

File: core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java [hunk 8]
```diff
@@ -198,11 +201,26 @@ public void upgradeFormatVersion() {
     assertTableUUID(requirements);
   }
 
+  @ParameterizedTest
+  @ValueSource(ints = {2, 3})
+  public void upgradeFormatVersionForView(int formatVersion) {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java:206]: actually we don't need to adjust any of the tests for views and this test can be removed, because views are currently at format-version = 2

File: core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java [hunk 9]
```diff
@@ -198,11 +201,26 @@ public void upgradeFormatVersion() {
...
+    assertThat(requirements)
+        .hasSize(1)
+        .hasOnlyElementsOfType(UpdateRequirement.AssertViewUUID.class);
+
+    assertViewUUID(requirements);
+  }
+
   @Test
-  public void upgradeFormatVersionForView() {
+  public void upgradeFormatVersionForViewV3() {
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java:220]: see my comment above, we can revert these changes here

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 10]
```diff
@@ -23,56 +23,100 @@
...
+    TableMetadata newTableMetadata = base.upgradeToFormatVersion(newTableVersion);
+
+    assertThat(
+            newTableMetadata.changes().stream()
+                .filter(
+                    metadataUpdate -> metadataUpdate instanceof MetadataUpdate.UpgradeFormatVersion)
+                .map(
+                    metadataUpdate ->
+                        ((MetadataUpdate.UpgradeFormatVersion) metadataUpdate).formatVersion()))
+        .isEqualTo(List.of(newTableVersion));
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:52]:
```suggestion
        .containsExactly(newFormatVersion);
```

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 11]
```diff
@@ -23,56 +23,100 @@
...
+    int newTableVersion = TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION;
+
+    TableMetadata newTableMetadata = base.upgradeToFormatVersion(newTableVersion);
+
+    // check that non-incremental updates are syntactic sugar for serial updates. E.g. upgrading
+    // from V1 to V3 will
+    // register changes in the table metadata for upgrading to V2 and V3 in order (V1->V2->V3)
+    assertThat(
+            newTableMetadata.changes().stream()
+                .filter(
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:73]: .filter(MetadataUpdate.UpgradeFormatVersion.class::isInstance)
        .map(MetadataUpdate.UpgradeFormatVersion.class::cast)
        .map(MetadataUpdate.UpgradeFormatVersion::formatVersion)

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 12]
```diff
@@ -23,56 +23,100 @@
...
+    return Arrays.asList(1, 2);
   }
 
   @TestTemplate
-  public void testDefaultFormatVersion() {
-    assertThat(table.ops().current().formatVersion()).isEqualTo(1);
+  public void testFormatVersionUpgrade() {
+    TableOperations ops = table.ops();
+    TableMetadata base = ops.current();
+    int baseTableVersion = base.formatVersion();
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:40]: actually there is `formatVersion` which carries the current version that is being set by `parameters()`. That way we don't need to keep track of `base` or `baseTableVersion`

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 13]
```diff
@@ -23,56 +23,100 @@
...
-    assertThat(ops.current().formatVersion()).isEqualTo(2);
+    assertThat(ops.current().formatVersion()).isEqualTo(newTableVersion);
   }
 
   @TestTemplate
   public void testFormatVersionUpgradeNotSupported() {
     TableOperations ops = table.ops();
     TableMetadata base = ops.current();
+    int baseTableVersion = base.formatVersion();
+    int unsupportedTableVersion = TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION + 1;
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:111]:
```suggestion
    int unsupportedFormatVersion = TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION + 1;
```

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 14]
```diff
@@ -23,56 +23,100 @@
...
+    assertThatThrownBy(() -> ops.commit(base, base.upgradeToFormatVersion(unsupportedTableVersion)))
         .isInstanceOf(IllegalArgumentException.class)
-        .hasMessage("Cannot upgrade table to unsupported format version: v4 (supported: v3)");
+        .hasMessage(
+            String.format(
+                "Cannot upgrade table to unsupported format version: v%d (supported: v%d)",
+                unsupportedTableVersion, TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION));
 
-    assertThat(ops.current().formatVersion()).isEqualTo(1);
+    assertThat(ops.current().formatVersion()).isEqualTo(baseTableVersion);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:120]:
```suggestion
    assertThat(ops.current().formatVersion()).isEqualTo(formatVersion);
```

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 15]
```diff
@@ -23,56 +23,100 @@
...
-    ops.commit(base, base.upgradeToFormatVersion(2));
+    int baseTableVersion = base.formatVersion();
+    int newTableVersion = baseTableVersion + 1;
+    ops.commit(base, base.upgradeToFormatVersion(newTableVersion));
 
-    assertThat(ops.current().formatVersion()).isEqualTo(2);
+    assertThat(ops.current().formatVersion()).isEqualTo(newTableVersion);
 
-    assertThatThrownBy(() -> ops.current().upgradeToFormatVersion(1))
+    assertThatThrownBy(() -> ops.current().upgradeToFormatVersion(baseTableVersion))
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:98]:
```suggestion
    assertThatThrownBy(() -> ops.current().upgradeToFormatVersion(formatVersion))
```

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 13 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:111]: replacing all instances of TableVersion with Format Version

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 12 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:40]: Purging baseTableVersion

Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:40]: Purging base

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 11 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:73]: so much cleaner thanks!

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 2 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java]: Added back

File: core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java (hunk 9 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java:220]: Missed that one. Removing

File: core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java (hunk 8 above)
Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestUpdateRequirements.java:206]: Makes sense, reverting the test

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 16]
```diff
@@ -1004,24 +1004,28 @@ private Builder setInitialFormatVersion(int newFormatVersion) {
...
-      if (newFormatVersion == formatVersion) {
+      if (targetFormatVersion == formatVersion) {
         return this;
       }
 
-      this.formatVersion = newFormatVersion;
-      changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      // register incremental version changes separately to ensure all upgrade requirements are met
+      for (int version = formatVersion + 1; version <= targetFormatVersion; version++) {
+        changes.add(new MetadataUpdate.UpgradeFormatVersion(version));
```
Comment (amogh-jahagirdar) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: More fundamental question (which I don't think we really need to answer in this PR, but maybe just think through) cc @RussellSpitzer @leangjonathan  @rdblue @nastra :

What if a REST server gets a list of upgrades out of order, like "1->3->2". I know here in the reference implementation that's not happening because of the loop but some other implementation theoretically could send changes in that order.

Do we need to change the REST spec to require a certain failure for this case?

I think ultimately we *do not* need to require that servers have a single upgrade path.

Servers can handle the upgrade path according to what they think is best (either by re-ordering those if they want to do the upgrade iteratively or by just doing the direct upgrade path if they like). This keeps the spec small and implementations can do whatever intelligent things they like.

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java (hunk 2 above)
Comment (amogh-jahagirdar) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:37]: Yeah it's really testing the formatVersion API which we should keep to make sure behavior stays the same no matter how many version there are!

**Review by amogh-jahagirdar — COMMENTED** [review]
Overall looks pretty good, just a naming nit and a question I think we want to think through but is not a blocker for this PR. Thank you @leangjonathan!

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 17]
```diff
@@ -1004,24 +1004,28 @@ private Builder setInitialFormatVersion(int newFormatVersion) {
       return this;
     }
 
-    public Builder upgradeFormatVersion(int newFormatVersion) {
+    public Builder upgradeFormatVersion(int targetFormatVersion) {
```
[the remark this answers is not in the document]
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1007]: That's fair. KISS everywhere it's possible :)

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 16 above)
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: +1

I was poking around the HTTP client, REST (de)serializers, and request parser logic. There is no content enforcement JSON inputs for deserialization . Specifically, the implementation of `UpdateTableRequestParser::fromJson` does not include any moderation of contents. Though I suppose deserialization and parsing wouldn't be the right place to enforce this so we come back to @nastra's concern about the REST spec and getting out of order or missing versions in the upgrade list.

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java (hunk 6 above)
Comment (amogh-jahagirdar) [reply core/src/test/java/org/apache/iceberg/TestTableMetadata.java:1473]: @leangjonathan just double checking since it was marked as resolved but I don't think it was.

**Review by amogh-jahagirdar — APPROVED** [review]
Overall LGTM, you'll probably want to double check if https://github.com/apache/iceberg/pull/10861/files#r1705024826 was meant to be resolved since I don't think the name change actually was performed. Thank you for the contribution, great to make progress here!

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 18]
```diff
@@ -1020,8 +1020,12 @@ public Builder upgradeFormatVersion(int newFormatVersion) {
         return this;
       }
 
+      // register incremental version changes separately to ensure all upgrade requirements are met
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:1023]: nit: I'd move this entire code block right after `this.formatVersion = newFormatVersion;`

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 16 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: Is there currently any downside in doing an upgrade from 1 to 3 directly without having an internal upgrade to 2? I believe the only restriction we have is if you downgrade the version, but there aren't any other restrictions that would prevent us from going 1 -> 3 directly?
Going 1 -> 3 direclty means that we don't have to create multiple changes of `MetadataUpdate.UpgradeFormatVersion` and not worry about a client potentially sending such an upgrade path out-of-order

Comment (amogh-jahagirdar) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: @nastra Check out the thread https://github.com/apache/iceberg/pull/10861#discussion_r1703406710, I had a similar question earlier but the rationale for the internal stepwise upgrade is to avoid having multiple upgrade paths from a table metadata builder perspective. 

So the the rationale for performing the stepwise upgrade at least from a client side perspective in the reference implementation (the change here):

The stepwise approach has the advantage of keeping testing a bit simpler (don't need to test 1->3 direct and 1->2->3, there's only a single path from 1 to 3) . It also seems better from a TableMetadata API perspective to ensure that any potential issues on the upgrade path have a consistent failure point. So if there's an issue upgrading from 2->3 in the 1->2->3 path, it's easy to distinguish that compared to 1->3 somehow just working (as an example).


I think in the end for the server side we can't really require a single upgrade path because the client can send any list of upgrade options, nor do I think we want to spec that out since we probably want to enable servers to make their choice? If they want to do the stepwise internal upgrade themselves since they consider it more safe or if they want to direct upgrade they can just do that.

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 18 above)
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1023]: We are iterating in the loop from `this.formatVersion` to start so we'd need to store it in another local var which is unnecessary IMO.

I did find a more concise way to write the loop though
```
      // register incremental version changes separately to ensure all upgrade requirements are met
      do {
        changes.add(new MetadataUpdate.UpgradeFormatVersion(++formatVersion));
      } while (formatVersion < newFormatVersion);

      this.formatVersion = newFormatVersion;
```

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 19]
```diff
@@ -1020,8 +1020,11 @@ public Builder upgradeFormatVersion(int newFormatVersion) {
         return this;
       }
 
-      this.formatVersion = newFormatVersion;
-      changes.add(new MetadataUpdate.UpgradeFormatVersion(newFormatVersion));
+      // register incremental version changes separately to ensure all upgrade requirements are met.
+      // formatVersion will be equivalent to newFormatVersion after the loop completes
+      do {
```
[the remark this answers is not in the document]
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: I can set it back to the original loop logic then

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 18 above)
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1023]: Reverted back to the original logic when this comment was made

File: core/src/main/java/org/apache/iceberg/TableMetadata.java (hunk 16 above)
Comment (leangjonathan) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: The problem we are solving involves how we maintain different upgrade paths. Currently, this stepwise internal upgrade is for future proofing version requirements. @amogh-jahagirdar @RussellSpitzer correct me if I'm wrong.

In the logic for upgrade requirements we currently have the following pattern:
```
  public static List<UpdateRequirement> forUpdateTable(
      TableMetadata base, List<MetadataUpdate> metadataUpdates) {
    Preconditions.checkArgument(null != base, "Invalid table metadata: null");
    Preconditions.checkArgument(null != metadataUpdates, "Invalid metadata updates: null");
    Builder builder = new Builder(base, false);
    builder.require(new UpdateRequirement.AssertTableUUID(base.uuid()));
    metadataUpdates.forEach(builder::update);
    return builder.build();
  }
```
Where `metadataUpdates.forEach(builder::update);` will register UpdateRequirements based on the specific metadata update into a list for later processing.

Using the 1, 2, 3 example...

Let's say that V2 required some field XYZ to be populated and we register that as an `UpdateRequirement` subclass.
```
  class AssertXYZPopulated implements UpdateRequirement {
    ...
  }
```
And we also fill in the appropriate builder logic to link `MetadataUpdate.UpgradeFormatVersion(2)` to `AssertXYZPopulated`

For V1->V2 upgrades the original logic worked since we would have a `MetadataUpdate.UpgradeFormatVersion(2)` instance registered and can thus enforce `AssertXYZPopulated`.

However, for V1->V3 upgrades, the original logic would have only had `MetadataUpdate.UpgradeFormatVersion(3)` (which does not have the mapping to `AssertXYZPopulated`) and we would fail to enforce the V2 upgrade requirements.

By adding the stepwise logic we are ensuring future upgrade requirements are always enforced version by version.

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: @leangjonathan I think the stepwise can be implemented on the upgrade itself rather than in our test framework. So if a user does Update 1 -> 3 on the back end we can do 1 -> 2 -> 3 so we necessarily would go through all the steps. I think we are ok just doing

N -> Latest and N -> N+1

The N+1 just for helping us track down stepwise pieces that are broken. Although technically if all the n -> n+1 upgrades work the n -> latest should also work if we implement it correctly.

Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: @nastra Do you have any other comment on this? I think we should get this and the other test stuff in asap so we can work on actual V3 Spec features

Comment (nastra) [reply core/src/main/java/org/apache/iceberg/TableMetadata.java:1025]: I've reached out to @rdblue to also get his opinion on any potential issues by doing a direct 1 -> 3 upgrade or whether we indeed would need to go through 1 -> 2 -> 3. 

The summary is that we should support a direct 1 -> 3 upgrade path for the following reason:
Basically any upgrade of the format version depends on whether it is backwards compatible. Currently, a V2 table is fully backwards compatible with V1 since all V1 metadata is valid in V2. 
Following the same approach, V2 metadata should most likely be valid in V3 which implies that V1 metadata would be valid in V3 as well and thus a direct 1 -> 3 should be supported

File: core/src/test/java/org/apache/iceberg/TestFormatVersions.java [hunk 20]
```diff
@@ -23,56 +23,95 @@
...
+  public void testFormatVersionUpgradeToLatest() {
+    TableOperations ops = table.ops();
+    int newFormatVersion = TableMetadata.SUPPORTED_TABLE_FORMAT_VERSION;
+
+    TableMetadata newTableMetadata = ops.current().upgradeToFormatVersion(newFormatVersion);
+
+    // check that non-incremental updates are syntactic sugar for serial updates. E.g. upgrading
+    // from V1 to V3 will register changes in the table metadata for upgrading to V2 and V3 in
+    // order (V1->V2->V3)
+    assertThat(
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/TestFormatVersions.java:70]: based on my last comment in https://github.com/apache/iceberg/pull/10861/files#r1718122028 around supporting a direct 1 -> 3 upgrade I believe you only need to remove this check here (or adjust it to only verify that there's a single `MetadataUpdate.UpgradeFormatVersion` instance if you prefer that). Given that the new format version in this test always points to 3, everything else in this test is still valid, so you don't really need to change much in terms of testing

Comment (nastra) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:70]: you might need to apply the same for `testFormatVersionUpgrade` but all the other tests in this class LGTM otherwise

Comment (leangjonathan) [reply core/src/test/java/org/apache/iceberg/TestFormatVersions.java:70]: Thanks! Reverting the changes in TableMetadata and I'll change the test to validate X -> X+1 and X -> latest

**Review by RussellSpitzer — APPROVED** [review]

**Comment by RussellSpitzer** [comment]
Thanks @leangjonathan ! Also thanks to @amogh-jahagirdar and @nastra for reviewing! We are one step closer to V3 :)
